### DISCUSSION TITLE
Using `eslint-plugin-react-hooks` together with `eslint-plugin-react-compiler`

### DISCUSSION THREAD

**Discussion by atomrevival** [reactwg/react-compiler#18] [General Questions]
Hi all! Just started experimenting with the React Compiler and was wondering if the rules for exhaustive deps are correct when using the compiler? 

I assumed that it wasn't necessary to use `useCallback` anymore on a function as a dependency, but the `eslint-plugin-react-hooks` plugin doesn't seem to agree? (Trying out the RC version `5.1.0-rc-8269d55d-20240802` as well just to see if I missed anything). Should I use this plugin at all when using the compiler?

Given the code below:

```ts
// ESLint: The 'myFunction' function makes the dependencies of useEffect Hook change on every render. 
// To fix this, wrap the definition of 'myFunction' in its own useCallback() Hook.
// (react-hooks/exhaustive-deps)
const myFunction = () => {
   // do something
};

useEffect(() => {
   myFunction();
}, [myFunction]); // Function as dependency
```

**Comment by poteto** [comment]
Sorry about the late response! Yeah, you can disable the exhaustive-deps rule if you're using the compiler's linter. The rules of hooks eslint rule is still helpful to have, although in most cases the compiler's linter should catch them too. 

It's a little duplicative right now but we are planning on merging these plugins in the near future. Thanks for the feedback!

**Reply by ahkhanjani** [reply]
> Sorry about the late response! Yeah, you can disable the exhaustive-deps rule if you're using the compiler's linter. The rules of hooks eslint rule is still helpful to have, although in most cases the compiler's linter should catch them too.

Coming from https://github.com/facebook/react/issues/31475. Shouldn't this be documented somewhere for the time being? The react compiler page in the official docs could include this

**Comment by Ganbin** [comment]
The issue I have with disabling the `exhaustive-deps` rule, is that it will correctly remove such error message: 

> [!CAUTION]
> The 'defaultValues' object makes the dependencies of useEffect Hook (at line 130) change on every render. To fix this, wrap the initialization of 'defaultValues' in its own useMemo() Hook. eslint([react-hooks/exhaustive-deps](https://github.com/facebook/react/issues/14920))

But it will fail to catch missing dependencies for our `useEffect` like : 
> [!CAUTION]
> React Hook useEffect has a missing dependency: 'form'. Either include it or remove the dependency array. eslint([react-hooks/exhaustive-deps](https://github.com/facebook/react/issues/14920))

I think the issue is that the rule is doing 2 different things. I hope the `react-compiler` linter will be able to show only the second error message. As of today, if I disable the `react-hooks/exhaustive-deps` the `react-compiler` linter doesn't catch the missing `useEffect` dependencies.

**Reply by Ganbin** [reply]
Any news on this issue? I just tried the latest 6.0.0-rc.1 and we still have the same issue as described in my previous post.

**Reply by BearCooder** [reply]
I think you might have a better to get a response chance by opening a bug in the reacr repo under the eslint-plugin-react-compiler category

**Reply by isaac-jordan** [reply]
We're treating this as blocking our adoption of react-compiler, as we don't want to miss out on having the linting rules for useEffect.

**Comment by alicerocheman** [comment]
Hi,
I'm also struggling with this issue, @poteto do you have any idea of a timeline for these plugins to be merge / the exhaustive deps issue to be resolved?
Thanks

**Comment by TravisJRyan** [comment]
Hi @poteto a related question on this (apologies if this is already documented, but I could not find clarity in the docs):

If I disable `react-hooks/exhaustive-deps` the react-compiler linter will error and claim the compiler won't run because the line was disabled. If I don't disable this and let it display as a warning, the react-compiler linter won't show anything, but it's unclear if it's still running.

I am working in a codebase that uses dependency arrays unidiomatically at times, and while I know we should fix that, I want to see if I need to fix all of those for the compiler to start adding value.

More generally, is there a way to have some visibility/tracing as to which components the compiler skipped (or is the react-compiler linter expected to be 100% reliable for this at the moment, even with the `react-hooks` linter also enabled)

**Comment by ChrisWiles** [comment]
Has this been fixed seems like it is still a issue

**Reply by freeboub** [reply]
I just tested 19.1.0-rc.2, I think this is still an issue.
What is strange is that compiler linter use/check useEffect, where in my knowledge, it doesn't optimize them  ...

**Reply by joaodematte** [reply]
Just a heads-up, the [official docs](https://react.dev/learn/react-compiler/installation#eslint-integration) recommend using the `eslint-plugin-react-hooks@rc` (without the `eslint-plugin-react-compiler`, I guess) version for better integration with the React Compiler's recommended settings. Might be worth checking out if you haven't already.

**Reply by uiii** [reply]
@joaodematte Thanks I tried it, unfortunately it doesn't help. At least I use the official recommended setup.

**Reply by weichensw** [reply]
This is still a issue as of today with every packages on latest version and using recommended configurations.

**Comment by ahkhanjani** [comment]
My experience with React Compiler has shown that you shouldn't trust the compiler fully. It can still definitely cause infinite render loops. If you're using something in useEffect, just use useMemo and useCallback (even if you need to apply a chain of things depending on each other). From what I know we will soon get automatic dependency array in the compiler in the future. That will get rid of the array entirely. Until then, don't ignore your lint warnings and don't risk an infinite loop in your apps. The compiler is not always that smart.

**Comment by sam6321** [comment]
Something else I've noticed that's related to this is getting `react-hooks/exhaustive-deps` and react compiler to play nice regarding custom hooks that should accept dependency arrays.

A hook with a signature such as this, for example

```tsx
function useCustomHook(callback: () => void, deps: unknown[]) {
    // ...
    useEffect(() => {
        // do something with callback in here
    }, deps); // <-- linter error occurs here
}
```

This will result in a linter error/warning on the deps line of the `useEffect`:
```
warning  React Hook useEffect was passed a dependency list that is not an array literal. This means we can't statically verify whether you've passed the correct dependencies  react-hooks/exhaustive-deps
```

For `react-hooks/exhaustive-deps`, this hook can be checked for dependencies by adding it to the `additionalHooks` option in eslint config:
```json
{
    "rules": {
        "react-hooks/exhaustive-deps": ["warn", { "additionalHooks": "useCustomHook" }],
    }
}
```

Then, the inner `useEffect` can have the eslint warning safely suppressed, since the deps check will occur correctly on the outer call to `useCustomHook` rather than the inner `useEffect` call:

```tsx
function useCustomHook(callback: () => void, deps: unknown[]) {
    // ...
    useEffect(() => {
        // do something with callback in here
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, deps);
}
```

This will now work correctly for eslint, but React Compiler will see the `// eslint-disable-next-line react-hooks/exhaustive-deps` and error out with messages similar to the following:

```
Compilation failed: useCustomHook.tsx
React Compiler has skipped optimizing this component because one or more React ESLint rules were disabled
React Compiler only works when your components follow all the rules of React, disabling them may result in unexpected or incorrect behavior. Found suppression `eslint-disable-next-line react-hooks/exhaustive-deps`
Suggestions: 'Remove the ESLint suppression and address the React error'
```

Basically, it seems like you can't make them both happy at the same time. Either eslint displays a warning, or react compiler does.

The obvious question that comes from this is, how can you write a custom hook that accepts dependencies that will allow both `react-hooks/exhaustive-deps` and react compiler to work properly?

I see two possibilities:
- React compiler needs to be given `"additionalHooks"` as well, and not error out on seeing the `// eslint-disable-next-line react-hooks/exhaustive-deps` inside of a component listed in `"additonalHooks"`.
- Or, `react-hooks/exhaustive-deps` needs to be smarter and identify when deps are used correctly inside of custom hook that is defined in `"additionalHooks"` and not generate a warning.

**Reply by kachkaev** [reply]
I guess you can suppress the warning by adding [`"use no memo"`](https://react.dev/reference/react-compiler/directives/use-no-memo) directive to your custom hook. Not sure cases like this can be compiled. What optimisations do you expect by the compilation in your case?

**Reply by sam6321** [reply]
Ideally, I'd expect react compiler to be able to perform the same optimisations as it would if the custom hook code was pasted directly into a component body. Suddenly de-optimising because you happened to move code into a custom hook function would not be ideal.

In my case though, I don't think there were any optimisations to perform and it was more about avoiding the react compiler error. `"use no memo"` is perfect for that, thank you.

**Reply by ahkhanjani** [reply]
> ```tsx
> function useCustomHook(callback: () => void, deps: unknown[]) {
>     // ...
>     useEffect(() => {
>         // do something with callback in here
>     }, deps); // <-- linter error occurs here
> }
> ```

This pattern looks pretty cursed ngl

**Reply by sam6321** [reply]
Could you explain why exactly?

This was the hook I was using:

```tsx
function useAsync<T>(callback: (abortSignal: AbortSignal) => Promise<T>, deps: unknown[] = []) {
    const [loading, setLoading] = useState(false);
    const [error, setError] = useState<Error | undefined>(undefined);
    const [result, setResult] = useState<T | undefined>(undefined);

    const innerCallback = useEffectEvent(callback);

    useEffect(() => {
        const abortController = new AbortController();
        setLoading(true);
        setError(undefined);

        (async () => {
            try {
                const res = await innerCallback(abortController.signal);
                abortController.signal.throwIfAborted();
                setResult(res);
            } catch (err) {
                if (!isAbortError(err)) {
                    setError(err as Error);
                }
            } finally {
                if (!abortController.signal.aborted) {
                    setLoading(false);
                }
            }
        })();

        return () => {
            abortController.abort();
        };

        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, deps);

    return { loading, error, result };
}
```

A straightforward async implementation that does not re-run on the callback changing, but does re-run on the deps changing.

Usage example:

```tsx
const { loading, error, result } = useAsync(async signal => {
    const r = await fetch(`some_resource/${someProp}`, {signal});
    return await r.json();
}, [someProp]);
```

**Reply by ahkhanjani** [reply]
> Could you explain why exactly?

ESLint and React Compiler need a static array. The compiler needs to know what causes rerenders at compile time. It doesn't have a runtime to adapt to changes in the dependency array.

> A straightforward async implementation that does not re-run on the callback changing, but does re-run on the deps changing.

You should move the useEffect responsible for firing your callback on state change out of the hook itself so you can handle it explicitly. Here's how I would rewrite your code:

```ts
function useAsync<T>(callback: (abortSignal: AbortSignal) => Promise<T>) {
  const [loading, setLoading] = useState(false);
  const [error, setError] = useState<Error | undefined>(undefined);
  const [result, setResult] = useState<T | undefined>(undefined);
  // controlled runs
  // shouldRunRef is there to prevent from firing on initial render
  // makes sure that the callback only runs when you explicitly call run()
  const shouldRunRef = useRef(false);
  const [runVersion, setRunVersion] = useState(0);

  const runEvent = useEffectEvent(async (signal: AbortSignal) => {
    setLoading(true);
    setError(undefined);

    try {
      // runs the latest version of the callback
      const res = await callback(signal);
      signal.throwIfAborted();
      setResult(res);
    } catch (err) {
      if (!isAbortError(err)) {
        setError(err as Error);
      }
    }
    // not using finally cuz React Compiler doesn't support it yet
    // should be fine
    if (!signal.aborted) {
      setLoading(false);
    }
  });

  // runs when you explicitly call run() from outside
  useEffect(() => {
    if (!shouldRunRef.current) {
      return;
    }
    const abortController = new AbortController();
    void runEvent(abortController.signal);

    return () => {
      abortController.abort();
    };
  }, [runVersion]);

  const run = useCallback(() => {
    shouldRunRef.current = true;
    setRunVersion((v) => v + 1);
  }, []);

  // return run to use it outside
  return { run, loading, error, result };
}
```

Now you can do this:

```ts
const { run, loading, error, result } = useAsync(async signal => {
    const r = await fetch(`some_resource/${someProp}`, {signal});
    return await r.json();
}); // no deps array here ❌

useEffect(() => {
  run();
}, [run, someProp]); // moved here and is now static ✅
```

**Reply by phryneas** [reply]
Why the deps array in the first place? Shouldn't `deps` influence the stability of `callback` instead of being passed separately?

It feels like if `deps` are not enough to guarantee `useCallback` stability of `callback`, they wouldn't work as correct deps for `useEffect` eiher.

```js
function useCustomHook(callback: () => void) {
  // ...
  useEffect(() => {
    // do something with callback in here
  }, [callback]);
}

// in a component
useCustomHook(
  useCallback(() => {
    // ...
  }, [dep1, dep2, etc])
);
```

**Reply by sam6321** [reply]
Yep, this is an obvious solution that also works, however it does introduce inconsistencies between different hook interfaces. Would a developer necessarily know that their `callback` needs to be stable?

It would be extremely easy for someone to do this and not realise:
```tsx
useCustomHook(async () => {
    // do something in here
});
```

Where as if you require them to pass a deps array, it makes it explicit that the hook re-renders on dep changes (and it also gets correctly linted by react-hooks/exhaustive-deps when added to additionalHooks)
```tsx
useCustomHook(async () => {
    // do something in here
}, []); // deps required by function, also matches useEffect and useCallback
```

**Reply by phryneas** [reply]
If they are using the compiler, that `useCallback` is implicit, though - they won't have to write it.
